ui/app/src/components/FormsEngine/lib/validators.ts

```diff
@@ -38,15 +39,53 @@ export const validatorsMap: Partial<Record<BuiltInControlType, ValidatorFunction
 	'file-name': undefined,
 	forcehttps: undefined,
 	'image-picker': undefined,
-	input: undefined,
+	input: (field, currentValue, messages) => {
+		let isValid = true;
+		// Skip validation if value is empty and field is not required
+		if (currentValue == null || (typeof currentValue === 'string' && currentValue.trim() === '')) {
+			return isValid;
+		}
+		const pattern = field.validations.pattern?.value as string;
+		// If there's a pattern and it doesn't match, it's invalid.
+		if (pattern && !String(currentValue).match(pattern)) {
+			messages.push(defineMessage({ defaultMessage: 'The value does not match the required pattern.' }));
+			isValid = false;
+		}
+		return isValid;
+	},
 	'internal-name': undefined,
 	label: undefined,
 	'link-input': undefined,
 	'link-textarea': undefined,
 	'linked-dropdown': undefined,
 	'locale-selector': undefined,
 	'node-selector': undefined,
-	'numeric-input': undefined,
+	'numeric-input': (field, currentValue, messages) => {
+		let isValid = true;
+		const pattern = field.validations.pattern?.value as string;
+		const maxValue = field.validations.maxValue?.value;
+		const minValue = field.validations.minValue?.value;
+
+		if (nnou(currentValue)) {
+			// If there's a pattern and it doesn't match
+			if (pattern && !String(currentValue).match(pattern)) {
+				messages.push(defineMessage({ defaultMessage: 'The value does not match the required pattern.' }));
+				isValid = false;
+			}
+			// If there's a max and the value is greater than the max
+			if (maxValue != null && Number(currentValue) > Number(maxValue)) {
+				messages.push(defineMessage({ defaultMessage: `The value is greater than the maximum allowed.` }));
```

Comment:
Add max to the message.

We could do something like:

```suggestion
				messages.push([defineMessage({ defaultMessage: `The value is greater than the allowed maximum ({maxValue}.)` }), { maxValue }]);
```

On the other end, if message is an array, assume `[descriptor, valuesObj]`